File: aegis/src/aegis/core/stream_names.py
"""Canonical CustomEvent names — the single source of truth shared by every module.

AG-UI carries domain payloads via ``CustomEvent(name, value)``. These constants are the
agreed ``name`` strings; the console mirrors them in ``web/src/lib/streamNames.ts``.
"""

from __future__ import annotations

REASONING = "reasoning"
GUARDRAIL_VERDICT = "guardrail_verdict"
#: An injection-classifier cache event (hit / miss) — the guardrail cost/observability feed.
GUARDRAIL_CACHE = "guardrail_cache"
#: A media guardrail verdict (image/audio): payload metadata plus the itemised list of
#: which media rails ran and which did not, so the console can never imply coverage
#: a control did not actually provide.
GUARDRAIL_MEDIA = "guardrail_media"
SHAP_EXPLANATION = "shap_explanation"
CONFORMAL_INTERVAL = "conformal_interval"
#: An ML model-card event — the honest, measured metadata of the model that is
#: serving (ensemble members, target/features, conformal coverage, split sizes).
ML_MODEL = "ml_model"
RETRIEVAL_CITATIONS = "retrieval_citations"
#: A retrieval semantic-cache event (hit / miss), carrying the cache provenance
#: (near-exact vs semantic, original query, cached-at) so the UI can show it.
RETRIEVAL_CACHE = "retrieval_cache"
ROUTING = "routing"
MEMORY_RECALL = "memory_recall"
#: A durable memory write (add/update/invalidate/delete/expire) — the memory changelog feed.
MEMORY_WRITE = "memory_write"
#: A semantic-cache event for recall (hit / miss / evict), with backend + similarity.
MEMORY_CACHE = "memory_cache"
MODEL_CALL = "model_call"
EVAL_RESULT = "eval_result"
#: The LLM-Ops self-improvement loop feed (aegis.ops). A Reflexion diagnose draft
#: (draft id + rationale + failure breakdown + risk tier).
OPS_DIAGNOSE = "ops_diagnose"
#: The release eval-gate + change-risk verdict (eval delta vs margin, risk tier, verdict).
OPS_GATE_DECISION = "ops_gate_decision"
#: The release outcome (promoted | staged_for_approval | rejected, version id, eval delta).
OPS_RELEASE = "ops_release"

ALL: frozenset[str] = frozenset(
    {REASONING, GUARDRAIL_VERDICT, GUARDRAIL_CACHE, GUARDRAIL_MEDIA, SHAP_EXPLANATION,
     CONFORMAL_INTERVAL,
     ML_MODEL, RETRIEVAL_CITATIONS, RETRIEVAL_CACHE, ROUTING, MEMORY_RECALL, MEMORY_WRITE,
     MEMORY_CACHE, MODEL_CALL, EVAL_RESULT, OPS_DIAGNOSE, OPS_GATE_DECISION, OPS_RELEASE}
)


def is_known(name: str) -> bool:
    """Return whether ``name`` is a registered CustomEvent name."""
    return name in ALL

File: web/src/lib/streamNames.ts
// Canonical CustomEvent names — mirrors aegis/src/aegis/core/stream_names.py `ALL` EXACTLY.
//
// AG-UI carries domain payloads via CustomEvent(name, value); these constants are the
// agreed `name` strings shared between the backend and this frontend mirror. Every entry
// in the Python `ALL` frozenset MUST appear here.
//
// Parity is enforced by `backend/tests/api/test_stream_name_mirror.py`, which parses THIS
// file and diffs it against the Python frozenset. That test exists because the previous
// guarantee was a comment claiming "parity is asserted by the count below" next to
// `STREAM_NAME_COUNT = STREAM_NAME_SET.size` — a count derived from this list, compared
// against itself. It could never fail, and while it sat here the mirror silently lost the
// five media/voice/vision names added by later modules. A check that cannot fail is worse
// than no check: it stops anyone from writing the real one.
export const STREAM_NAMES = {
  REASONING: 'reasoning',
  GUARDRAIL_VERDICT: 'guardrail_verdict',
  GUARDRAIL_CACHE: 'guardrail_cache',
  GUARDRAIL_MEDIA: 'guardrail_media',
  SHAP_EXPLANATION: 'shap_explanation',
  CONFORMAL_INTERVAL: 'conformal_interval',
  ML_MODEL: 'ml_model',
  RETRIEVAL_CITATIONS: 'retrieval_citations',
  RETRIEVAL_CACHE: 'retrieval_cache',
  ROUTING: 'routing',
  MEMORY_RECALL: 'memory_recall',
  MEMORY_WRITE: 'memory_write',
  MEMORY_CACHE: 'memory_cache',
  MODEL_CALL: 'model_call',
  EVAL_RESULT: 'eval_result',
  OPS_DIAGNOSE: 'ops_diagnose',
  OPS_GATE_DECISION: 'ops_gate_decision',
  OPS_RELEASE: 'ops_release',
  VOICE_CHUNK: 'voice_chunk',
  VOICE_TRANSCRIPT: 'voice_transcript',
  VISION_SCREEN: 'vision_screen',
  VISION_ANALYSIS: 'vision_analysis',
  WEB_SEARCH: 'web_search',
} as const

/** The canonical set of every known stream name (mirrors the Python `ALL`). */
export const STREAM_NAME_SET: ReadonlySet<string> = new Set(Object.values(STREAM_NAMES))

/** Number of known stream names — must equal len(aegis.core.stream_names.ALL). */
export const STREAM_NAME_COUNT = STREAM_NAME_SET.size

/** Return whether `name` is a registered CustomEvent name (mirrors is_known). */
export function isKnownStreamName(name: string): boolean {
  return STREAM_NAME_SET.has(name)
}

export type StreamName = (typeof STREAM_NAMES)[keyof typeof STREAM_NAMES]
